src/nimble.nim

```diff
@@ -1539,30 +1588,25 @@ proc lock(options: Options) =
   ## it if it already exists.
 
   let currentDir = getCurrentDir()
-  let pkgInfo = getPkgInfo(currentDir, options)
+  var pkgInfo = getPkgInfo(currentDir, options)
 
   let doesLockFileExist = displayLockOperationStart(currentDir)
   validateDevModeDepsWorkingCopiesBeforeLock(pkgInfo, options)
 
-  let dependencies = pkgInfo.processFreeDependencies(options).map(
-    pkg => pkg.toFullInfo(options)).toSeq
+  let
+    packageList = getInstalledPkgsMin(options.getPkgsDir(), options)
+    developDeps = processDevelopDependencies(pkgInfo, options)
+    installedInfo = installIteration(concat(packageList, developDeps), pkgInfo.requires, options)
+    dependencies =
+      block:
+        # To avoid warning on `map`
+        var res = toSeq(installedInfo.installed.values)
+        res.apply(x => x.toFullInfo(options))
+        res
+    dependenciesGraph = dependencies.buildLockFileDeps(installedInfo.dependencies, options)
+
   pkgInfo.validateDevelopDependenciesVersionRanges(dependencies, options)
-  var dependencyGraph = buildDependencyGraph(dependencies, options)
-
-  if currentDir.lockFileExists:
-    # If we already have a lock file, merge its data with the newly generated
-    # one.
-    #
-    # IMPORTANT TODO:
-    # To do this properly, an SMT solver is needed, but anyway, it seems that
-    # currently Nimble does not check properly for `require` clauses
-    # satisfaction between all packages, but just greedily picks the best
-    # matching version of dependencies for the currently processed package.
-
-    dependencyGraph = mergeLockedDependencies(pkgInfo, dependencyGraph, options)
-
-  let (topologicalOrder, _) = topologicalSort(dependencyGraph)
-  writeLockFile(dependencyGraph, topologicalOrder)
+  writeLockFile(dependenciesGraph, toSeq(dependenciesGraph.keys()).sorted)
```

Comment:
I modified this bit because it was based on the `PackageInfo.requires`, but the requires is less useful than the dependency graph returned by `installIteration`.
The returned dependency graph will be fully resolved (eg, in require there could be `https://github.com/status-im/nim-unittest2.git` which will become `unittest2` in the dependency graph)

I'm also not sure why the merge system was there? Instead this version will generate a predictable lockfile (since everything is sorted), so if one dependency change, only the lines of the dependency will be impacted.